src/NHibernate/Linq/ReWriters/RemoveUnnecessaryBodyOperators.cs

```diff
@@ -21,20 +22,32 @@ public static void ReWrite(QueryModel queryModel)
 
 		internal static void RemoveUnnecessaryOrderByClauses(QueryModel queryModel)
 		{
-			if (queryModel.ResultOperators.Count == 1 &&
-			    queryModel.ResultOperators.All(
-				    r => r is ContainsResultOperator || r is AnyResultOperator || r is AllResultOperator))
+			if (IsOrderByNeeded(queryModel))
+				return;
+
+			// For these operators, we can remove any order-by clause
+			var bodyClauses = queryModel.BodyClauses;
+			for (int i = bodyClauses.Count - 1; i >= 0; i--)
 			{
-				// For these operators, we can remove any order-by clause
-				var bodyClauses = queryModel.BodyClauses;
-				for (int i = bodyClauses.Count - 1; i >= 0; i--)
-				{
-					if (bodyClauses[i] is OrderByClause)
-						bodyClauses.RemoveAt(i);
-				}
+				if (bodyClauses[i] is OrderByClause)
+					bodyClauses.RemoveAt(i);
 			}
 		}
 
+		internal static bool IsOrderByNeeded(QueryModel queryModel)
+		{
+			switch (queryModel.ResultOperators.Count)
+			{
+				case 1:
+					var r = queryModel.ResultOperators[0];
+					return !(r is AnyResultOperator || r is AllResultOperator || r is ContainsResultOperator);  
+				case 0:
```

Comment:
This should probably go into the aggregate paging rewiter instead.

Reply:
meaning?

Reply:
I tried to say that this logic probably belongs to some other rewriter.

Reply:
I don't get  how do you see it expressed in code? 
Something like:
```C#
case 0:
return AggregatingGroupJoinRewriter.IsOrderByNeeded(queryModel):
```

Reply:
No.

Reply:
I was thinking of moving the entire `0` case into `AggregatingGroupJoinRewriter` where we create `NhAggregatedExpression`'s. But it did not work, because this check is also used when we copy order-by into the `QueryModel` with `NhAggregatedExpression`